File: packages/server/src/transport.ts
import type { ServerModuleTransport } from "./launcher.ts";

export interface InMemoryServerModuleTransport extends ServerModuleTransport {
  listSubjects(): string[];
  request(subject: string, payload: unknown): Promise<unknown>;
}

export function createInMemoryServerModuleTransport(): InMemoryServerModuleTransport {
  const handlers = new Map<string, (payload: unknown) => Promise<unknown>>();

  return {
    async subscribe(subject: string, handler: (payload: unknown) => Promise<unknown>) {
      handlers.set(subject, handler);
    },
    listSubjects() {
      return [...handlers.keys()].sort((left, right) => left.localeCompare(right));
    },
    async request(subject: string, payload: unknown) {
      const handler = handlers.get(subject);
      if (!handler) {
        throw new Error(`missing handler for ${subject}`);
      }

      return await handler(payload);
    },
  };
}

File: crates/bladb-server/src/transport.rs
use serde_json::Value;
use std::{
    collections::HashMap,
    sync::{Arc, Mutex},
};

pub type TransportHandler = Arc<dyn Fn(Value) -> Result<Value, String> + Send + Sync + 'static>;

pub trait ServerModuleTransport: Send + Sync {
    fn subscribe(&self, subject: String, handler: TransportHandler) -> Result<(), String>;
}

#[derive(Clone, Default)]
pub struct InMemoryServerModuleTransport {
    handlers: Arc<Mutex<HashMap<String, TransportHandler>>>,
}

impl InMemoryServerModuleTransport {
    pub fn list_subjects(&self) -> Vec<String> {
        let Ok(handlers) = self.handlers.lock() else {
            return vec![];
        };
        let mut subjects = handlers.keys().cloned().collect::<Vec<_>>();
        subjects.sort();
        subjects
    }

    pub fn request(&self, subject: &str, payload: Value) -> Result<Value, String> {
        let handler = self
            .handlers
            .lock()
            .map_err(|_| "server module transport lock poisoned".to_string())?
            .get(subject)
            .cloned()
            .ok_or_else(|| format!("missing handler for {subject}"))?;

        handler(payload)
    }
}

impl ServerModuleTransport for InMemoryServerModuleTransport {
    fn subscribe(&self, subject: String, handler: TransportHandler) -> Result<(), String> {
        self.handlers
            .lock()
            .map_err(|_| "server module transport lock poisoned".to_string())?
            .insert(subject, handler);
        Ok(())
    }
}

pub fn create_in_memory_server_module_transport() -> InMemoryServerModuleTransport {
    InMemoryServerModuleTransport::default()
}
